File: ios/Packages/HapiKit/Sources/HapiClient/SSE/ReconnectPolicy.swift
import Foundation

/// Normative SSE timing constants, in milliseconds.
///
/// Values come from `docs/api/client-contract/sse.md` ("Reconnect policy")
/// and the web reference client (`web/src/hooks/useSSE.ts`).
public enum SSETimings {
    /// The hub emits a heartbeat frame every 30 s (informational).
    public static let heartbeatIntervalMs = 30_000
    /// No frames of any kind for 90 s ⇒ tear down and reconnect.
    public static let stalenessThresholdMs = 90_000
    /// On app-foreground, reconnect immediately when the last frame is older
    /// than this. One missed heartbeat interval after an OS suspend is
    /// already enough to distrust the socket (no FIN/RST ever surfaces).
    public static let foregroundResumeStalenessMs = 45_000
    /// Staleness check interval. Checks are skipped while suspended.
    public static let watchdogTickMs = 10_000
    /// An attempt that has not reached OPEN within 10 s is likely hung on a
    /// dead pooled socket — abandon it and retry on a fresh connection.
    public static let connectTimeoutMs = 10_000
}

/// Pure backoff schedule for SSE reconnects.
///
/// Delay for attempt `n` (0-based, reset to 0 on every successful open):
///
/// - attempt 0 → 0 ms (the first retry is immediate; backoff is for
///   *repeated* failures, not the initial recovery),
/// - attempt n ≥ 1 → `min(cap, base × 2^(n-1))`,
/// - cap is 30 s, widening to 300 s once `n ≥ 8` (a hub that stays
///   unreachable is usually down for hours, and every retry through a relay
///   costs a TLS handshake),
/// - plus uniform jitter of 0–500 ms on every delay, immediate ones included.
///
/// Jitter is injectable: the generic `delayMs(forAttempt:using:)` draws from
/// any `RandomNumberGenerator` so tests can seed it, and callers that manage
/// randomness themselves can add `exponentialDelayMs(forAttempt:)` + their
/// own draw from `jitterRangeMs` (this is what `SSEClient` does).
public struct ReconnectPolicy: Equatable, Sendable {
    public var baseDelayMs: Int
    public var maxDelayMs: Int
    public var slowMaxDelayMs: Int
    public var slowAfterAttempts: Int
    public var jitterRangeMs: ClosedRange<Int>

    public init(
        baseDelayMs: Int = 1_000,
        maxDelayMs: Int = 30_000,
        slowMaxDelayMs: Int = 300_000,
        slowAfterAttempts: Int = 8,
        jitterRangeMs: ClosedRange<Int> = 0...500
    ) {
        self.baseDelayMs = baseDelayMs
        self.maxDelayMs = maxDelayMs
        self.slowMaxDelayMs = slowMaxDelayMs
        self.slowAfterAttempts = slowAfterAttempts
        self.jitterRangeMs = jitterRangeMs
    }

    /// The deterministic part of the schedule, without jitter.
    public func exponentialDelayMs(forAttempt attempt: Int) -> Int {
        guard attempt >= 1 else {
            return 0
        }
        let cap = attempt >= slowAfterAttempts ? slowMaxDelayMs : maxDelayMs
        // Clamp the exponent so the shift can never overflow; 2^30 × base is
        // already far past every cap.
        let exponent = min(attempt - 1, 30)
        return min(cap, baseDelayMs << exponent)
    }

    /// Full delay including a jitter draw from the supplied generator.
    public func delayMs<R: RandomNumberGenerator>(forAttempt attempt: Int, using rng: inout R) -> Int {
        exponentialDelayMs(forAttempt: attempt) + Int.random(in: jitterRangeMs, using: &rng)
    }
}

File: android/core/data/src/main/kotlin/app/hapi/data/sse/ReconnectPolicy.kt
package app.hapi.data.sse

import kotlin.random.Random

/**
 * The normative SSE reconnect constants and backoff schedule from
 * `docs/api/client-contract/sse.md` ("Reconnect policy") / the web reference
 * client (`web/src/hooks/useSSE.ts`). Pure — all randomness is injected so
 * tests can pin the exact schedule.
 */
object ReconnectPolicy {
    /** An attempt that has not completed the handshake in 10 s is hung — abandon it. */
    const val CONNECT_TIMEOUT_MS: Long = 10_000

    /** No frames of any kind (heartbeat included) for 90 s ⇒ tear down and reconnect. */
    const val STALE_MS: Long = 90_000

    /**
     * On app-foreground, a connection whose last frame is older than 45 s is
     * reconnected immediately: an OS suspend can kill the socket without any
     * error surfacing, and one missed 30 s heartbeat already means distrust.
     */
    const val RESUME_STALE_MS: Long = 45_000

    /** Staleness check interval. Checks are skipped while backgrounded. */
    const val WATCHDOG_TICK_MS: Long = 10_000

    const val BASE_DELAY_MS: Long = 1_000
    const val MAX_DELAY_MS: Long = 30_000
    /** Uniform 0..[JITTER_MS] (inclusive) added to every delay. */
    const val JITTER_MS: Int = 500
    /** After this many consecutive failures the ceiling widens to [SLOW_MAX_DELAY_MS]. */
    const val SLOW_AFTER_ATTEMPTS: Int = 8
    /** A hub unreachable for 8 straight attempts is usually down for hours. */
    const val SLOW_MAX_DELAY_MS: Long = 300_000

    /**
     * Delay before retry number `attempt` (0-based count of consecutive
     * failures so far, reset to 0 on every successful handshake):
     *
     * - attempt 0 → 0 ms exponential part (first retry is immediate, jitter only);
     * - attempt n ≥ 1 → `min(cap, 1000 · 2^(n-1))`, where the cap is 30 s and
     *   widens to 300 s once `n ≥ 8`;
     * - plus uniform jitter 0..500 ms on every delay.
     *
     * Schedule (jitter aside): 0, 1s, 2s, 4s, 8s, 16s, 30s, 30s, 128s, 256s,
     * 300s, 300s, …
     */
    fun delayForAttempt(attempt: Int, random: Random): Long {
        require(attempt >= 0) { "attempt must be >= 0, got $attempt" }
        val cap = if (attempt >= SLOW_AFTER_ATTEMPTS) SLOW_MAX_DELAY_MS else MAX_DELAY_MS
        val exponential = when {
            attempt == 0 -> 0L
            // 2^(n-1) would overflow / always exceed the cap long before n=63.
            attempt - 1 >= 20 -> cap
            else -> minOf(cap, BASE_DELAY_MS shl (attempt - 1))
        }
        return exponential + random.nextInt(JITTER_MS + 1)
    }
}
